### PR TITLE
Fix #5910 - Spoof ICMP echo replies

### PR DESCRIPTION
**Pull request by jetvova**
**Description**
Some apps check network connectivity by sending ICMP pings. Before this fix, ICMP traffic was swallowed by mitmproxy_rs, which makes them believe that there is no network connectivity. 

Mitmproxy now generates fake ICMP Echo Replies in response to ICMP Echo Requests, and transmits them back to the sender's address.

### CODE REVIEW

**Comment by jetvova** [comment]
Rewrote the implementation to use a simpler approach without going all the way to python, as advised.

**Review by mhils — COMMENTED** [review]
This looks great, thanks. Could you please add tests in https://github.com/mitmproxy/mitmproxy_rs/blob/main/src/network/tests.rs? Otherwise LGTM. :)

File: src/network/task.rs [hunk 1]
```diff
@@ -220,6 +224,30 @@ impl<'a> NetworkIO<'a> {
...
+        // swallowed by mitmproxy_rs, which makes them believe that there is no network connectivity.
+        // Generating fake ICMP replies as a simple workaround.
+
+        let permit = match self.net_tx.try_reserve() {
+            Ok(p) => p,
+            Err(_) => {
+                log::debug!("Channel full, discarding ICMP packet.");
+                return Ok(());
+            }
+        };
```
[the remark this answers is not in the document]
Comment (jetvova) [reply src/network/task.rs]: Done.

**Comment by mhils** [comment]
I manually pulled this into main and fixed the newlines issues. Thanks again! 🍰 😃

**Pull request closed without merging**